src/rnnlm/arpa-sampling-test.cc

```diff
@@ -0,0 +1,216 @@
+// lm/arpa-sampling-test.cc
+
+#include <math.h>
+#include <typeinfo>
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "fst/fstlib.h"
+#include "arpa-sampling.h"
+
+namespace kaldi {
+
+class ArpaSamplingTest {
+ public:
+  typedef ArpaSampling::HistType HistType;
+  typedef ArpaSampling::WordToProbsMap WordToProbsMap;
+  typedef ArpaSampling::NgramType NgramType;
+  typedef ArpaSampling::HistWeightsType HistWeightsType;
+  
+  ArpaSamplingTest (ArpaSampling *arpa) {
+    arpa_ = arpa;
+  } 
+  // This function reads in a list of histories and their weights from a file
+  // only text form is supported
+  void ReadHistories(std::istream &is, bool binary, 
+      std::vector<std::pair<HistType, BaseFloat> > *histories);
+  
+  // This function tests the correctness of the read-in ARPA LM 
+  void TestReadingModel();
+
+  // This function tests the generated unigram distribution 
+  void TestUnigramDistribution();
+
+  // Test non-unigram words and alpha
+  // TODO: need to test alpha in a proper way 
+  void TestGetDistribution(const
+      std::vector<std::pair<HistType, BaseFloat> > &histories);
+ private:
+  // This ArpaSampling object is used to get accesses to private and protected
+  // members in ArpaSampling class
+  ArpaSampling *arpa_;
+};
+
+void ArpaSamplingTest::ReadHistories(std::istream &is, bool binary,
+    std::vector<std::pair<HistType, BaseFloat> > *histories) {
+  if (binary) {
+    KALDI_ERR << "binary-mode reading is not implemented for ArpaFileParser";
+  }
+  const fst::SymbolTable* sym = arpa_->Symbols();
+  std::string line;
+  KALDI_LOG << "Start reading histories from file...";
+  // int32 ngram_order = arpa_->ngram_order_;
+  int32 ngram_order = arpa_->ngram_order_;
+  while (getline(is, line)) {
+    std::istringstream is(line);
+    std::istream_iterator<std::string> begin(is), end;
+    std::vector<std::string> tokens(begin, end);
+    HistType history;
+    int32 word;
+    BaseFloat hist_weight = 0;
+    for (int32 i = 0; i < tokens.size() - 1; ++i) {
+      word = sym->Find(tokens[i]);
+      if (word == fst::SymbolTable::kNoSymbol) {
+        word = sym->Find(arpa_->unk_symbol_);
```

Comment:
You can remove this.  We won't be doing anything special about words not in the vocabulary; they would just be an error.